pkg/configuration/configuration.go

```diff
@@ -484,9 +575,87 @@ func (ev *extendedViper) GetAllKeysThatContainValues(key string) []string {
 }
 
 func (ev *extendedViper) GetKeyType(key string) KeyType {
-	if strings.HasPrefix(strings.ToLower(key), "snyk_") {
-		return EnvVarKeyType
+	// check for supported env vars
+	for _, envVar := range ev.supportedEnvVars {
+		if strings.EqualFold(key, envVar) {
+			return EnvVarKeyType
+		}
+	}
+
+	// check for supported prefixes
+	for _, prefix := range ev.supportedEnvVarPrefixes {
+		if strings.HasPrefix(strings.ToLower(key), strings.ToLower(prefix)) {
+			return EnvVarKeyType
+		}
 	}
 
 	return UnspecifiedKeyType
 }
+
+// AutomaticEnv wraps Viper's AutomaticEnv and allows us to check if this has been set or not.
+// If AutomaticEnv is enabled, SetSupportedEnvVars and SetSupportedEnvVarPrefixes is disabled.
+func (ev *extendedViper) AutomaticEnv() {
+	ev.mutex.Lock()
+	defer ev.mutex.Unlock()
+
+	ev.viper.AutomaticEnv()
+	ev.automaticEnvEnabled = true
+}
+
+func (ev *extendedViper) SetAutomaticEnv(b bool) {
+	ev.automaticEnvEnabled = b
+}
+
+func (ev *extendedViper) GetAutomaticEnv() bool {
+	return ev.automaticEnvEnabled
+}
+
+// SetSupportedEnvVars sets the env vars that will be supported regardless of prefixes
+// Only needed when SetAutomaticEnv is not used
+func (ev *extendedViper) SetSupportedEnvVars(envVars ...string) {
+	ev.mutex.Lock()
+	defer ev.mutex.Unlock()
+
+	for _, envVar := range envVars {
+		if slices.Contains(ev.supportedEnvVars, envVar) {
+			return
+		}
+
+		ev.supportedEnvVars = append(ev.supportedEnvVars, envVar)
+	}
+}
+
+func (ev *extendedViper) GetSupportedEnvVars() []string {
+	return ev.supportedEnvVars
+}
+
+// SetSupportedEnvVarPrefixes sets prefixes which env vars must have to be supported
+// Only needed when SetAutomaticEnv is not used
+func (ev *extendedViper) SetSupportedEnvVarPrefixes(prefixes ...string) {
+	ev.mutex.Lock()
+	defer ev.mutex.Unlock()
+
+	for _, prefix := range prefixes {
+		if slices.Contains(ev.supportedEnvVarPrefixes, prefix) {
+			return
+		}
+
+		ev.supportedEnvVarPrefixes = append(ev.supportedEnvVarPrefixes, prefix)
+	}
+}
+
+func (ev *extendedViper) GetSupportedEnvVarPrefixes() []string {
+	return ev.supportedEnvVarPrefixes
+}
+
+// SetFiles sets the files from which the Configuration instance uses
+func (ev *extendedViper) SetFiles(files ...string) {
+	ev.mutex.Lock()
+	defer ev.mutex.Unlock()
+
+	ev.configFiles = append(ev.configFiles, files...)
+}
+
+func (ev *extendedViper) GetFiles() []string {
+	return ev.configFiles
```

Comment:
Suggestion: I know it is not consistent in the existing code, but you should be locking the mutex even in a getter.

Reply:
I used `ev.mutex.TryLock()` in these instances to avoid hitting a deadlock with the calling method's (`Clone()`) `mutex.Lock`. 

Open to alternatives, if `TryLock` is not appropriate here

Reply:
I think TryLock() is making the situation worse to be honest, as it'll fail if the mutex is already locked, since it is unknown who locked the mutex this effectively ignoring if someone else holds it. Which is exactly what we want to avoid using the mutex.

Some additional thoughts on solutions:
* The mutex we acquire during Clone() is supposed to ensure that the configuration to clone `ev` isn't changed while cloning.
* The deadlocks only happen when we use the getter of `ev`, possible solution access the fields directly without the locking public accessors.
* Use a [RWMutex](https://go.dev/pkg/sync/?m=old#RWMutex), which allows multiple reader but just a single writer.

From the complexity of the change I think avoiding getter in clone and still having all public methods use the mutex is the best.

Reply:
BTW, taking the trylock change as an proof that we are missing concurrency tests, let's add one that fails. We have -race enabled so that should be easily discoverable.